docs/plugin/framework: Initial validation page (#1967)
hashicorp/terraform-website: content/source/docs/plugin/framework/validation.html.md

---
layout: "extend"
page_title: "Plugin Development - Framework: Validation"
description: |-
  How to validate configuration values using the provider development framework.
---

# Validation

Practitioners implementing Terraform configurations desire feedback surrounding the syntax, types, and acceptable values. This feedback, typically referred to as validation, is preferably given as early as possible before a configuration is applied. Terraform supports validation for values in provider, resource, and data source configurations. The framework supports returning [diagnostics](./diagnostics.html) feedback on all of these, offering the ability to specify validations on an attribute, provider-defined type, and the entire provider, resource, or data source schema.

~> **NOTE:** When implementing validation logic, configuration values may be [unknown](./types.html#unknown) based on the source of the value. Implementations must account for this case, typically by returning early without returning new diagnostics.

## Syntax and Basic Schema Validation

> The [Terraform configuration language](https://www.terraform.io/docs/language/) is declarative and an implementation of [HashiCorp Configuration Language](https://github.com/hashicorp/hcl) (HCL). Terraform CLI is responsible for reading and parsing configurations for validity, based on Terraform's concepts such as `resource` blocks and associated syntax. Basic validation of value type and behavior information, for example returning an error when a string value is given where a list value is expected or returning an error when a required attribute is missing from a configuration, is automatically handled by Terraform CLI based on the provider, resource, or data source schema.

Reviewer: ```suggestion
The [Terraform configuration language](https://www.terraform.io/docs/language/) is declarative and an implementation of [HashiCorp Configuration Language](https://github.com/hashicorp/hcl) (HCL). The Terraform CLI is responsible for reading and parsing configurations for validity, based on Terraform's concepts such as `resource` blocks and associated syntax. Basic validation of value type and behavior information, for example returning an error when a string value is given where a list value is expected or returning an error when a required attribute is missing from a configuration, is automatically handled by the Terraform CLI based on the provider, resource, or data source schema.
```

Any further validation provided by the framework occurs after these checks.

## Attribute Validation

> It is common for provider implementations to introduce validation on attributes using the generic framework-defined types such as `types.String`. The [`tfsdk.Attribute` type `Validators` field](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#Attribute.Validators) can be supplied with a list of validations and diagnostics will be returned from all validators. For example:

Reviewer: ```suggestion
It is common for provider implementations to introduce validation on attributes using the generic framework-defined types such as [`types.String`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/types#String). The [`tfsdk.Attribute` type `Validators` field](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#Attribute.Validators) can be supplied with a list of validations and diagnostics will be returned from all validators. For example:
```

```go
// Typically within the tfsdk.Schema returned by GetSchema() for a provider,
// resource, or data source.
tfsdk.Attribute{
    // ... other Attribute configuration ...

    Validators: []AttributeValidators{
        // These are example validators
        stringLengthBetween(10, 256),
        stringRegularExpression(regexp.MustCompile(`^[a-z0-9]+$`)),
    },
}
```

> 

Reviewer: ```suggestion

All validators will always be run, regardless of whether previous validators returned an error or not.

```

-> The framework will implement or reference common use case attribute validations, such as string length, in the future.

### Creating Attribute Validators

To create an attribute validator, the [`tfsdk.AttributeValidator` interface](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#AttributeValidator) must be satisfied. For example:

```go
type stringLengthBetweenValidator struct {
    Max int
    Min int
}

> 

Reviewer: ```suggestion

// Description returns a plain text description of the validator's behavior, suitable for a practitioner to understand its impact.
```

func (v stringLengthBetweenValidator) Description(ctx context.Context) string {
    return fmt.Sprintf("string length must be between %d and %d", v.Min, v.Max)
}

> 

Reviewer: ```suggestion

// MarkdownDescription returns a markdown formatted description of the validator's behavior, suitable for a practitioner to understand its impact.
```

func (v stringLengthBetweenValidator) MarkdownDescription(ctx context.Context) string {
    return fmt.Sprintf("string length must be between `%d` and `%d`", v.Min, v.Max)
}

> 

Reviewer: ```suggestion

// Validate runs the main validation logic of the validator, reading configuration data out of `req` and updating `resp` with diagnostics.
```

func (v stringLengthBetweenValidator) Validate(ctx context.Context, req tfsdk.ValidateAttributeRequest, resp *tfsdk.ValidateAttributeResponse) {

>     str, ok := request.AttributeConfig.(types.String)
> 
>     if !ok {
>         response.Diagnostics.AddAttributeError(
>             request.AttributePath,
>             "Invalid Attribute Value Type",
>             fmt.Sprintf("Expected types.String value, got: %T", request.AttributeConfig),
>         )
> 
>         return
>     }

Reviewer: ```suggestion
    // types.String must be the attr.Value produced by the attr.Type in the schema for this attribute
    // for generic validators, use
    // https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#ConvertValue
    // to convert into a known type.
    var str types.String
    diags := tfsdk.ValueAs(ctx, val, &target)
    resp.Diagnostics.Append(diags...)
    if resp.Diagnostics.HasError() {
        return
    }
```

This explains that we can only use the `attr.Value` produced by the `attr.Type` in the schema, tells people how to convert between `attr.Type`s, and uses the `tfsdk.ValueAs` instead of type assertion to get the value.

Author: Love it, just needs minor variable name fixes.

```suggestion
    // types.String must be the attr.Value produced by the attr.Type in the schema for this attribute
    // for generic validators, use
    // https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#ConvertValue
    // to convert into a known type.
    var str types.String
    diags := tfsdk.ValueAs(ctx, req.AttributeConfig, &str)
    resp.Diagnostics.Append(diags...)
    if diags.HasError() {
        return
    }
```

    if str.Unknown || str.Null {
        return
    }

    strLen := len(str)

    if strLen < v.Min || strLen > v.Max {
        response.Diagnostics.AddAttributeError(
            request.AttributePath,

>             "Attribute Validation Error",

Reviewer: Can we be more descriptive here, to prompt provider devs to write better errors?

```suggestion
            "Invalid String Length",
```

>             fmt.Sprintf("%s, got: %d", v.Description(ctx), strLen),

Reviewer: Does this confuse the matter of what `Description` is for? Do we want to de-duplicate it to make it clear the point of Description is for documentation and not to be part of the validator?

Author: Let's "duplicate" for now to reduce any confusion (and correctly do more of sentence structure).

```suggestion
            fmt.Sprintf("String length must be between %d and %d, got: %d.", v.Min, v.Max, strLen),
```

---
layout: "extend"
page_title: "Plugin Development - Framework: Validation"
description: |-
  How to validate configuration values using the provider development framework.
---

# Validation

> Practitioners implementing Terraform configurations desire feedback surrounding the syntax, types, and acceptable values. This feedback, typically referred to as validation, is preferably given as early as possible before a configuration is applied. Terraform supports validation for values in provider, resource, and data source configurations. The framework supports returning [diagnostics](./diagnostics.html) feedback on all of these, offering the ability to specify validations on an attribute, provider-defined type, and the entire provider, resource, or data source schema.

Reviewer 2: Just thinking about how we can make this as concise as possible :) 

```suggestion
The framework can return [diagnostics](./diagnostics.html) feedback for values in provider, resource, and data source configurations. This allows you to write validations that give users feedback about required syntax, types, and acceptable values.
```

> ~> **NOTE:** When implementing validation logic, configuration values may be [unknown](./types.html#unknown) based on the source of the value. Implementations must account for this case, typically by returning early without returning new diagnostics.

Reviewer 2: Can we move this note down to where folks will be reading about implementation? I just want them to see it in context - I worry they'll forget about it if it stays up here :)

Author: Certainly makes sense! I'm worried about repeating this note up to 5 times (Creating Attribute Validators, Defining Type Validation, and the 3 schema validation sections). Do you have a suggestion for what do in this case?

Reviewer 2: fair - I will leave this note here then. :)

> ## Syntax and Basic Schema Validation

Reviewer 2: I want this heading to make the purpose clear - this is about when the checks you write with the framework occur, right? 

Also - just to be clear - if you write validation it will happen every single time we run one of the command you list below, right? So you write it one time and it gets applied everywhere.

Author: Newer practitioners and provider developers are generally curious about which type of validation is handled by which part of Terraform (CLI does some, SDK does some, provider does some). The general idea behind this section at a high level was to delineate between which of the validation responsibilities occurs where and when. I was thinking the general context might be helpful. It was maybe named poorly to highlight that this section is specifically what happens with Terraform CLI.

The Terraform CLI validation occurs before any framework/provider validation. Providers are not required to implement anything to receive the syntax and basic types validation. Maybe that's the key part missing from here that could help clarify this better.

Reviewer 2: yeah I trust you on this but I wonder if we should say idk "Built-In Validation" "Default Validation"  or something similar? or even just "Default Terraform CLI Validation"?  yay. titles :p 

I just think this section title feels vague and so I want to be more specific about what it's about

Author: Oh! I like Default Terraform CLI Validation -- let's give that a go. Thank you!

```suggestion
## Default Terraform CLI Validation
```

We can circle back about this after the ❄️ , if desired. 😄

> The [Terraform configuration language](https://www.terraform.io/docs/language/) is declarative and an implementation of [HashiCorp Configuration Language](https://github.com/hashicorp/hcl) (HCL). The Terraform CLI is responsible for reading and parsing configurations for validity, based on Terraform's concepts such as `resource` blocks and associated syntax. Basic validation of value type and behavior information, for example returning an error when a string value is given where a list value is expected or returning an error when a required attribute is missing from a configuration, is automatically handled by the Terraform CLI based on the provider, resource, or data source schema.

Reviewer 2: ```suggestion
The [Terraform configuration language](https://www.terraform.io/docs/language/) is declarative and an implementation of [HashiCorp Configuration Language](https://github.com/hashicorp/hcl) (HCL). The Terraform CLI is responsible for reading and parsing configurations for validity, based on Terraform's concepts such as `resource` blocks and associated syntax. The Terraform CLI automatically handles basic validation of value type and behavior information based on the provider, resource, or data source schema. For example, the Terraform CLI returns an error when a string value is given where a list value is expected and also when a required attribute is missing from a configuration.
```

> Any further validation provided by the framework occurs after these checks.

Author: ```suggestion
These syntax and basic schema checks are performed during the [`terraform apply`](https://www.terraform.io/docs/cli/commands/apply.html), [`terraform destroy`](https://www.terraform.io/docs/cli/commands/destroy.html), [`terraform plan`](https://www.terraform.io/docs/cli/commands/plan.html), and [`terraform validate`](https://www.terraform.io/docs/cli/commands/validate.html) commands. Any further validation provided by the framework occurs after these checks.
```

> These syntax and basic schema checks are performed during the [`terraform apply`](https://www.terraform.io/docs/cli/commands/apply.html), [`terraform destroy`](https://www.terraform.io/docs/cli/commands/destroy.html), [`terraform plan`](https://www.terraform.io/docs/cli/commands/plan.html), and [`terraform validate`](https://www.terraform.io/docs/cli/commands/validate.html) commands. Any further validation provided by the framework occurs after these checks.

Reviewer 2: ```suggestion
Terraform CLI syntax and basic schema checks occur during the [`terraform apply`](https://www.terraform.io/docs/cli/commands/apply.html), [`terraform destroy`](https://www.terraform.io/docs/cli/commands/destroy.html), [`terraform plan`](https://www.terraform.io/docs/cli/commands/plan.html), and [`terraform validate`](https://www.terraform.io/docs/cli/commands/validate.html) commands. Any additional validation you define with the framework occurs directly after these checks are complete.
```

## Attribute Validation

> It is common for provider implementations to introduce validation on attributes using the generic framework-defined types such as [`types.String`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/types#String). The [`tfsdk.Attribute` type `Validators` field](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#Attribute.Validators) can be supplied with a list of validations and diagnostics will be returned from all validators. For example:

Reviewer 2: ```suggestion
You can introduce validation on attributes using the generic framework-defined types such as [`types.String`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/types#String). To do this, supply the [`tfsdk.Attribute` type `Validators` field](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#Attribute.Validators) with a list of validations, and the framework will return diagnostics from all validators. For example:
```

```go
// Typically within the tfsdk.Schema returned by GetSchema() for a provider,
// resource, or data source.
tfsdk.Attribute{
    // ... other Attribute configuration ...

    Validators: []AttributeValidators{
        // These are example validators
        stringLengthBetween(10, 256),
        stringRegularExpression(regexp.MustCompile(`^[a-z0-9]+$`)),
    },
}
```

All validators will always be run, regardless of whether previous validators returned an error or not.

> 
> -> The framework will implement or reference common use case attribute validations, such as string length, in the future.

Author: ```suggestion
```

> -> The framework will implement or reference common use case attribute validations, such as string length, in the future.

Reviewer 2: We probably don't want to say this only because it's making promises and we should avoid forward-looking statements. Instead, we probably just want to say that the framework does not currently reference common use case attribute validations. OR we omit this entirely :)

Author: Let's omit it for now. 😄

### Creating Attribute Validators

> To create an attribute validator, the [`tfsdk.AttributeValidator` interface](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#AttributeValidator) must be satisfied. For example:

Reviewer 2: ```suggestion
To create an attribute validator, you must implement the [`tfsdk.AttributeValidator` interface](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#AttributeValidator). For example:
```

}

// MarkdownDescription returns a markdown formatted description of the validator's behavior, suitable for a practitioner to understand its impact.
func (v stringLengthBetweenValidator) MarkdownDescription(ctx context.Context) string {
    return fmt.Sprintf("string length must be between `%d` and `%d`", v.Min, v.Max)
}

// Validate runs the main validation logic of the validator, reading configuration data out of `req` and updating `resp` with diagnostics.
func (v stringLengthBetweenValidator) Validate(ctx context.Context, req tfsdk.ValidateAttributeRequest, resp *tfsdk.ValidateAttributeResponse) {
    // types.String must be the attr.Value produced by the attr.Type in the schema for this attribute
    // for generic validators, use
    // https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#ConvertValue
    // to convert into a known type.
    var str types.String
    diags := tfsdk.ValueAs(ctx, req.AttributeConfig, &str)
    resp.Diagnostics.Append(diags...)
    if diags.HasError() {
        return
    }

    if str.Unknown || str.Null {
        return
    }

    strLen := len(str)

    if strLen < v.Min || strLen > v.Max {

>         response.Diagnostics.AddAttributeError(
>             request.AttributePath,

Author: ```suggestion
        resp.Diagnostics.AddAttributeError(
            req.AttributePath,
```

            "Invalid String Length",
            fmt.Sprintf("String length must be between %d and %d, got: %d.", v.Min, v.Max, strLen),
        )

        return
    }
}
```

Optionally and depending on the complexity, it may be desirable to also create a helper function to instantiate the validator. For example:

```go
func stringLengthBetween(minLength int, maxLength int) stringLengthBetweenValidator {
    return stringLengthBetweenValidator{
        Max: maxLength,
        Min: minLength,
    }
}
```

## Type Validation

> Providers that contain common attribute values with consistent validation rules may wish to create a custom type to simplify schemas. When validation is implemented on a type, it is not necessary to declare the same validation on the attribute, although additional validations can be supplied in that manner. For example:

Reviewer 2: ```suggestion
You may want to create a custom type to simplify schemas if your provider that contains common attribute values with consistent validation rules. When you implement validation on a type, you do not need to declare the same validation on the attribute, but you can supply additional validations in that manner. For example:
```

Author: Just removing an extra word from the suggestion 😄 

```suggestion
You may want to create a custom type to simplify schemas if your provider contains common attribute values with consistent validation rules. When you implement validation on a type, you do not need to declare the same validation on the attribute, but you can supply additional validations in that manner. For example:
```

```go
// Typically within the tfsdk.Schema returned by GetSchema() for a provider,
// resource, or data source.
tfsdk.Attribute{
    // ... other Attribute configuration ...

    // This is an example type which implements its own validation
    Type: computeInstanceIdentifierType,

    // This is optional, example validation that is checked in addition
    // to any validation performed by the type
    Validators: []AttributeValidators{
        stringLengthBetween(10, 256),
    },
}
```

### Defining Type Validation

> To support validation within a type, the [`attr.TypeWithValidate` interface](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/attr#TypeWithValidate) must be satisfied. For example:

Reviewer 2: ```suggestion
To support validation within a type, you must implement the [`attr.TypeWithValidate` interface](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/attr#TypeWithValidate). For example:
```

type computeInstanceIdentifierType struct {}

func (t computeInstanceIdentifierType) Validate(ctx context.Context, tfValue tftypes.Value, path *tftypes.AttributePath) diag.Diagnostics {
    var diags diag.Diagnostics

    if !tfValue.Type().Equal(tftypes.String) {
        diags.AddAttributeError(
            path,
            "Compute Instance Type Validation Error",
            fmt.Sprintf("Expected String value, received %T with value: %v", in, in),
        )
        return diags
    }

    if !tfValue.IsKnown() || tfValue.IsNull() {
        return diags
    }

    var value string
    err := tfValue.As(&value)

    if err != nil {
        diags.AddAttributeError(
            path,
            "Compute Instance Type Validation Error",
            fmt.Sprintf("Cannot convert value to string: %s", err),
        )
        return diags
    }

    if !strings.HasPrefix(value, "instance-") {
        diags.AddAttributeError(
            path,
            "Compute Instance Type Validation Error",
            fmt.Sprintf("Missing `instance-` prefix, got: %s", value),
        )
        return diags
    }
}
```

## Schema Validation

> Provider, resource, and data source schemas also support validation across all attributes. This is helpful when checking values in multiple attributes, such as only accepting certain values in one attribute when another is a specific value.

Reviewer 2: Could we make the example after "such as" more specific/a real-world scenario?

Author: Maybe this will be slightly better (it is a little hard because these tend to be very-circumstance specific):

```suggestion
Provider, resource, and data source schemas also support validation across all attributes. This is helpful when checking values in multiple attributes, such as ensuring the values are compatible with each other.
```

### Creating Provider Schema Validation

> There are two possible interface implementations that can be used for provider validation. Either or both can be implemented. This validation is performed in addition to any attribute and type validation within the provider schema.

Reviewer 2: ```suggestion
The framework performs provider validation in addition to attribute and type validation. You can implement either or both of the following interfaces.
```

> The [`tfsdk.ProviderWithConfigValidators` interface](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#ProviderWithConfigValidators) follows a similar pattern to attribute validation and allows for a more declarative approach, which is helpful for consistent validators across multiple providers. Each of the validators must satify the [`tfsdk.ProviderConfigValidator` interface](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#ProviderConfigValidator). For example:

Reviewer 2: ```suggestion
The [`tfsdk.ProviderWithConfigValidators` interface](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#ProviderWithConfigValidators) follows a similar pattern to attribute validation and allows for a more declarative approach. This lets you write consistent validators across multiple providers. You must implement the [`tfsdk.ProviderConfigValidator` interface](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#ProviderConfigValidator) for each validator. For example:
```

```go
// Other methods to implement the tfsdk.Provider interface are omitted for brevity
type exampleProvider struct {}

func (p exampleProvider) ConfigValidators(ctx context.Context) []tfsdk.ProviderConfigValidator {
    return []tfsdk.ProviderConfigValidator{
        /* ... */
    }
}
```

> The [`tfsdk.ProviderWithValidateConfig` interface](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#ProviderWithValidateConfig) is more imperative in design and simplifies one-off functionality that typically applies to a single provider. For example:

Reviewer 2: ```suggestion
The [`tfsdk.ProviderWithValidateConfig` interface](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#ProviderWithValidateConfig) is more imperative in design and is useful for validating unique functionality that typically applies to a single provider. For example:
```

```go
// Other methods to implement the tfsdk.Provider interface are omitted for brevity
type exampleProvider struct {}

func (p exampleProvider) ValidateConfig(ctx context.Context, req ValidateProviderConfigRequest, resp *ValidateProviderConfigResponse) {
    // Retrieve values via req.Config.Get() or req.Config.GetAttribute(),
    // then return any warnings or errors via resp.Diagnostics.
}
```

### Creating Resource Schema Validation

> There are two possible interface implementations that can be used for resource validation. Either or both can be implemented. This validation is performed in addition to any attribute and type validation within the resource schema.

Reviewer 2: ```suggestion
The framework performs resource schema validation in addition to any attribute and type validation. You can implement either or both of the following interfaces.
```

> The [`tfsdk.ResourceWithConfigValidators` interface](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#ResourceWithConfigValidators) follows a similar pattern to attribute validation and allows for a more declarative approach, which is helpful for consistent validators across multiple resources. Each of the validators must satify the [`tfsdk.ResourceConfigValidator` interface](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#ResourceConfigValidator). For example:

Reviewer 2: ```suggestion
The [`tfsdk.ResourceWithConfigValidators` interface](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#ResourceWithConfigValidators) follows a similar pattern to attribute validation and allows for a more declarative approach. This lets you create consistent validators across multiple resources. You must implement the [`tfsdk.ResourceConfigValidator` interface](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#ResourceConfigValidator) for each validator. For example:
```

```go
// Other methods to implement the tfsdk.Resource interface are omitted for brevity
type exampleResource struct {}

func (r exampleResource) ConfigValidators(ctx context.Context) []tfsdk.ResourceConfigValidator {
    return []tfsdk.ResourceConfigValidator{
        /* ... */
    }
}
```

> The [`tfsdk.ResourceWithValidateConfig` interface](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#ResourceWithValidateConfig) is more imperative in design and simplifies one-off functionality that typically applies to a single resource. For example:

Reviewer 2: ```suggestion
The [`tfsdk.ResourceWithValidateConfig` interface](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#ResourceWithValidateConfig) is more imperative in design and is useful for validating unique functionality that typically applies to a single resource. For example:
```

```go
// Other methods to implement the tfsdk.Resource interface are omitted for brevity
type exampleResource struct {}

func (r exampleResource) ValidateConfig(ctx context.Context, req ValidateResourceConfigRequest, resp *ValidateResourceConfigResponse) {
    // Retrieve values via req.Config.Get() or req.Config.GetAttribute(),
    // then return any warnings or errors via resp.Diagnostics.
}
```

### Creating Data Source Schema Validation

> There are two possible interface implementations that can be used for data source validation. Either or both can be implemented. This validation is performed in addition to any attribute and type validation within the data source schema.

Reviewer 2: ```suggestion
The framework performs data source schema validation in addition to any attribute and type validation. You can implement either or both of the following interfaces.
```
